spatialprofilingtoolbox/db/feature_matrix_extractor.py

```diff
@@ -18,24 +18,61 @@
 
 class FeatureMatrixExtractor:
     """
-    Pull from the database and create convenience bundle of feature matrices
-    and metadata.
+    Pull from the database and create convenience bundle of feature matrices and metadata.
     """
-    @staticmethod
-    def extract(database_config_file, specimen: str=None, study: str=None, continuous_also=False):
-        E = FeatureMatrixExtractor
-        data_arrays = E.retrieve_expressions_from_database(database_config_file,
-                                                           specimen=specimen,
-                                                           study=study,
-                                                           continuous_also=continuous_also)
-        centroid_coordinates = E.retrieve_structure_centroids_from_database(database_config_file,
-                                                                            specimen=specimen,
-                                                                            study=study)
-        stratification = E.retrieve_derivative_stratification_from_database(database_config_file)
-        study_component_lookup = E.retrieve_study_component_lookup(database_config_file)
-        merged = E.merge_dictionaries(
-            E.create_feature_matrices(data_arrays, centroid_coordinates),
-            E.create_channel_information(data_arrays),
+
+    cursor: Psycopg2Cursor | None
+    database_config_file: str | None
+
+    def __init__(self,
+        cursor: Psycopg2Cursor | None=None,
+        database_config_file: str | None=None,
+    ):
+        if cursor is None and database_config_file is None:
+            logger.error('Must supply either cursor or database_config_file.')
+            return
+        if cursor is not None and database_config_file is not None:
+            message = 'A cursor and database configuration file were both specified. Using the '\
+                'cursor.'
+            logger.warning(message)
+        self.cursor = cursor
+        self.database_config_file = database_config_file
```

Comment:
The warning says that the cursor will be used if both cursor and config are provided, but it saves both anyway.

Reply:
Yes I went back and forth with a few options here.
Originally I wanted the cursor setup to take place immediately, here, with no need for a `database_config_file` attribute. But the context manager business frustrated this, as usual.
The expected behavior is for `cursor` to get used if provided, otherwise use `database_config_file` if available.
It is helpful to store `database_config_file` in `self.database_config_file` even when it is `None`, because it tells that no `database_config_file` was supplied. To make this transparent we could have some additional state variable like `db_source: DBSource`, with enum `DBSource` members `CURSOR` or `CONFIG_FILE`. But that seems like overkill.

Reply:
Well overkill or not I implemented it, and it seems cleaner.